### PR TITLE
core/state: concurrent storage and account trie updates

### PR DESCRIPTION
**Pull request by karalabe**
Blocked by https://github.com/ethereum/go-ethereum/pull/19953.

### CODE REVIEW

File: core/state/statedb.go [hunk 1]
```diff
@@ -634,6 +635,47 @@ func (self *StateDB) GetRefund() uint64 {
...
+		end   time.Time
+		lock  sync.Mutex
+	)
+	for _, obj := range updated {
+		obj := obj // Closure, take care
+		workers.Schedule(func() {
+			obj.updateRoot(s.db)
+			if metrics.EnabledExpensive {
+				lock.Lock()
+				end = time.Now()
```
Comment (holiman) [inline core/state/statedb.go]: Perhaps use an atomic here instead?

File: core/state/workers.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import "runtime"
+
+// workers is a singleton pool of goroutines to run arbitrary tasks concurrently.
+var workers = newWorkerPool(16 * runtime.GOMAXPROCS(0))
```
Comment (holiman) [inline core/state/workers.go]: Why `16 x ` the number of CPUs? Why not just `newWorkerPool(runtime.GOMAXPROCS(0))` ?

Comment (holiman) [reply core/state/workers.go]: Ah, so the pool still only uses `runtime.GOMAXPROCS` channels, but the queue has some slack so we don't have to wait for a thread to be ready.

File: core/state/statedb.go (hunk 1 above)
Comment (karalabe) [reply core/state/statedb.go]: Yeah, it's pretty ugly I agree. Trying to clean the code up a bit and get rid of some suboptimal calls in the mean time.

File: core/state/workers.go (hunk 2 above)
Comment (karalabe) [reply core/state/workers.go]: Yeah, that was the rationale. Another alternative could be to make the pool even fancier and support a Join operation. So each "source" can queue up tasks (maybe unlimited), but then they could also join the pool in crunching through them (only own tasks) to ensure that if the pool is busy on another task (e.g. prefetcher, rpc?, tracer), then the main thread doesn't wait, rather progresses guaranteed. Not sure, only ideas here.

**Comment by karalabe** [comment]
Although this idea might have merit, it's a bit complex, and the behavior will change drastically if we add snapshotting (it might make more sense there). Bottom line is that this PR as is is probably not valuable enough on top of master, and will only serve to make things harder to bechmark wrt snapshotting. Lets "ship" that first if good enough and revisit concurrency afterwards.

**Pull request closed without merging**